### PR TITLE
SONARJAVA-2408 S4143 Map values should not be replaced unconditionally

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/OverwrittenKeyCheck.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+import org.sonar.plugins.java.api.tree.ExpressionStatementTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import javax.annotation.Nullable;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/OverwrittenKeyCheck.java]: MINOR Remove this unused import 'javax.annotation.Nullable'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: java-checks/src/main/java/org/sonar/java/checks/OverwrittenKeyCheck.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+@Rule(key = "S4143")
+public class OverwrittenKeyCheck extends IssuableSubscriptionVisitor {
+
+  private static final MethodMatcher MAP_PUT = MethodMatcher.create().typeDefinition("java.util.Map").name("put").parameters(TypeCriteria.anyType(), TypeCriteria.anyType());
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/OverwrittenKeyCheck.java]: TypeDefinition should be `TypeCriteria.subtypeOf("java.util.Map")`, or you are never going to catch stuff like:
```
  HashMap<Object, Object> map2;
  void foo() {
    map2.put(2, "foo");
    map2.put(2, "bar"); // Noncompliant
  }
```

File: java-checks/src/main/java/org/sonar/java/checks/OverwrittenKeyCheck.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    return Collections.singletonList(Tree.Kind.BLOCK);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+
+    Map<CollectionAndKey, Tree> usedKeys = new HashMap<>();
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/OverwrittenKeyCheck.java]: It seems to me that you don't need a map here but you may use only a `set<CollectionAndKey>`. If there is an issue detected, you can use the trees from the new `CollectionAndKey` and the one already existing from the set.

File: java-checks/src/main/java/org/sonar/java/checks/OverwrittenKeyCheck.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        }
+      }
+    }
+    return null;
+  }
+
+  private void handleKey(Map<CollectionAndKey, Tree> keys, CollectionAndKey collectionAndKey) {
+    if (keys.containsKey(collectionAndKey)) {
+      Tree previousTree = keys.get(collectionAndKey);
+      String indexOrKey = collectionAndKey.keyTree.parent().is(Tree.Kind.ARRAY_DIMENSION) ? "index" : "key";
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/OverwrittenKeyCheck.java]: I would encode the type of the collection into the Object, and put a method there, or precompute it when creating the object.

File: java-checks/src/main/java/org/sonar/java/checks/OverwrittenKeyCheck.java (hunk 3 above)
Comment (saberduck) [reply java-checks/src/main/java/org/sonar/java/checks/OverwrittenKeyCheck.java]: As discussed, you are right that object in the Set would have the right tree, but there is no API to get the object which was inserted from the Set .

**Pull request merged**